modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
         return threadPool;
     }
 
-    static class Stats {
+    enum Operation {
+        GET_OBJECT("GetObject"),
+        LIST_OBJECTS("ListObjects"),
+        PUT_OBJECT("PutObject"),
+        PUT_MULTIPART_OBJECT("PutMultipartObject"),
+        DELETE_OBJECTS("DeleteObjects"),
+        ABORT_MULTIPART_OBJECT("AbortMultipartObject");
 
-        final AtomicLong listCount = new AtomicLong();
+        private final String value;
 
-        final AtomicLong getCount = new AtomicLong();
+        Operation(String value) {
+            this.value = value;
+        }
+    }
 
-        final AtomicLong putCount = new AtomicLong();
+    class StatsCollectors {
+        final ConcurrentMap<String, IgnoreNoResponseMetricsCollector> collectors;
+
+        StatsCollectors() {
+            this.collectors = Stream.of(
+                Operation.GET_OBJECT,
+                Operation.LIST_OBJECTS,
+                Operation.PUT_OBJECT,
+                Operation.PUT_MULTIPART_OBJECT,
+                Operation.DELETE_OBJECTS,
+                Operation.ABORT_MULTIPART_OBJECT
+            ).collect(Collectors.toConcurrentMap(ops -> ops.value, this::buildMetricCollector));
+        }
 
-        final AtomicLong postCount = new AtomicLong();
+        RequestMetricCollector getMetricCollector(Operation operation, OperationPurpose purpose) {
+            final String key;
+            if (purpose == OperationPurpose.SNAPSHOT) {
+                key = operation.value;
+            } else {
+                key = operation.value + "/" + purpose.getValue();
+                if (false == collectors.containsKey(key)) {
+                    collectors.put(key, buildMetricCollector(operation));
```

Comment:
Suggest using `putIfAbsent` so that this step is atomic.

Reply:
🤦 Right. Thanks!